lrpar/src/lib/lex_api.rs

```diff
@@ -5,15 +5,28 @@ use std::{cmp, error::Error, fmt, hash::Hash, marker};
 use cfgrammar::Span;
 use num_traits::{PrimInt, Unsigned};
 
+#[derive(Copy, Clone, Debug)]
+pub struct StartStateId {
+    id: usize,
+}
```

Comment:
So I'm now seeing there is a bit of difficulty because we're cross-crate issues, in getting a `StartState` here, from `StartStateId`. Compounded somewhat by *lrpar* not depending on *lrlex*.  I also noticed that `get_start_state_by_id` is in a trait `LexerDef` so seems to be implicitly `pub` already.

Just not sure how to deal with this situation where `LexError`/`StartStateId` are in *lrpar*, but `StartState` is in *lrlex*.
I'm going to have to think on it.

Reply:
Okay, I'll hold off on further code-churn while you consider!

Reply:
Well, I pushed a first attempt here which just makes a bunch of accessor functions for `LexError`/`StartStateId` https://github.com/ratmice/grmtools/tree/start_state_ids
Which seems to work i'm not sure it is super elegant, I hadn't really anticipated that we can't just use `StartStateId.id` directly inside lrlex.

I guess though, where I'm having trouble with this, is after the addition of these accessors + `LexError::new` + `StartState::try_from` with infallible errors, we don't actually gain much in the way invariants ensuring that `StartStateId`s are valid.

However I guess they still ensure that it is difficult to mix them up with arbitrary usizes?

Reply:
> Just not sure how to deal with this situation where LexError/StartStateId are in lrpar, but StartState is in lrlex.

Hmm, yeah, good point -- I hadn't thought about that.

What operations does a user need to do on a `StartState`? Is it just "compare two state states for equality and print it out for debugging purposes"? If so, we could e.g. make `LexError` something like:

```rust
pub struct LexError<StartStateId> {
    span: Span,
    lexing_state: Option<StartStateId>,
}
```

and then lrlex can return `LexError<StartStateId>` where it defines `struct StartStateId(usize)` or similar?

[Warning: this might be a terrible idea! It's off the top of my head.]

Reply:
It does look like adding that as a generic is not going to be a walk in the park, as it entails threading the generic throughout everything starting with the LexError within https://softdevteam.github.io/grmtools/master/api/lrpar/trait.Lexer.html

From there it spreads pretty heavily throughout all the other generics, in ways which I haven't succeeded in adding it to yet.

Reply:
Another thing we could do if we wanted is have a `trait StartStateId` and `lexing_state: Option<Box<dyn StartStateId>>` but we'd to know what functionality a `StartStateId` should have. If it's just `Eq + Debug + Display`, that should be easy?

Reply:
Well, part of the difficulty here is, you need a reference to the concrete `LRNonStreamingLexerDef`, to actually `Debug`/`Display` anything meaningful.  If we revisited putting `StartState` directly in `LexError`, we would have something meaningful to print for `Debug`/`Display` though, it might also want to be `Eq + Debug + Display + Spanned`.

E.g. the `Debug` impl we're after first requires a call to something like `fn (lex: lrlex::LrNonStreamingLexerDef, state: lrpar::StartStateId) -> lrlex::StartState`.  Because `StartStateId` is just kind of a internal implementation detail which only has meaning to the lexer itself, and not so much to the user.

Reply:
If it's purely internal, is there even any point exposing this to the user?

[In retrospect, I suspect the `Copy` on `LexError` was a mistake. I don't know how difficult removing it now might be though...]

Reply:
Yeah currently there is a point, in that we can get something relevant to the lex file from it.
But if we can just embed `StartState` directly, there is no point in exposing it.

When developing lex files with start states here, I've found it can be fairly confusing when you get a `LexError`, and a regex doesn't match the regex you expect, but not because the regex is wrong but because the lexer is in a different state than you expect.

So I think it is important to give a hint here regarding which state the lexer was in, because the typical reaction is that there is something wrong about the regex.

Reply:
OK, so then I guess the question is: should `StartStartId` just be a "token" (no pun intended) that you can hand back to the lexer to say "give me the name of the state you're in"? Because if that's all we have to do with it, we can probably hack it with a trait that's an `Any` (it's not pretty... but it would work). Either way, I think I'm inclined to get this PR in if we're happy with it, and then fix the `LexError` issue later. Any thoughts?

Reply:
That is kind of how https://github.com/ratmice/grmtools/commit/ca3444534b84a8dca736e9275b7bfe376bf9ce4c worked except `StartStateId` is concrete rather than `Any`, from there the name can be gotten by `lexer.error_lexing_state(err).map(|state| state.name())` or something to that effect.

So yes something along those lines of it being just a token seems like it should work fine.
I'd also be fine if we want to fix the `LexError` issue in another patch, and am happy with the state of things in this one.